core/modules/memman/MemFile.cc

```diff
@@ -107,6 +107,21 @@ MemFile::MLResult MemFile::memLock() {
         return nilResult;
     }
 
+    // TODO: Find a better solution for systems where mmap is not viable, which
+    // manifests as no bytes being locked but bytes are reserved.
+    // Fake mmap as the configuration isn't working properly.
+    // At this point, there is enough free space, it was not a flex table,
+    // but mmap failed. So,this will fake it being a flexilock file and
+    // reserve space for it.
+    if (_memInfo.size() < freeBytes) {
```

Comment:
After some thought, I think the appropriate change is to add "NEEDLOCK" 
attribute that would work like the modified code, intermediate between 
MUSTLOCK and FLEXLOCK.

Andy

On Fri, 29 Apr 2016, John Gates wrote:

>> @@ -107,6 +107,21 @@ MemFile::MLResult MemFile::memLock() {
>>          return nilResult;
>>      }
>>
>> +    // TODO: Find a better solution for systems where mmap is not viable, which
>> +    // manifests as no bytes being locked but bytes are reserved.
>> +    // Fake mmap as the configuration isn't working properly.
>> +    // At this point, there is enough free space, it was not a flex table,
>> +    // but mmap failed. So,this will fake it being a flexilock file and
>> +    // reserve space for it.
>> +    if (_memInfo.size() < freeBytes) {
>
> I'm not happy with it either. That's why it starts out with todo. However, without this change, memman doesn't work well on the cluster at all (queries on large tables go from 6 hours to 8 hours). I think we will need something like this for use on systems where the workers don't have enough memory and/or can't be easily configured so that mmap works.
>
> ---
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/lsst/qserv/pull/220/files/856bf25bb7aed922d9dc8ea78cb71835ec3bbaf2#r61603150